lasagne/layers/shape.py

```diff
@@ -139,6 +139,62 @@ def get_output_for(self, input, *args, **kwargs):
 reshape = ReshapeLayer # shortcut
 
 
+class DimShuffleLayer(Layer):
+    def __init__(self, incoming, pattern):
+        super(DimShuffleLayer, self).__init__(incoming)
+
+        # Sanity check the pattern
+        used_dims = set()
+        for p in pattern:
+            if isinstance(p, int):
+                # Dimension p
+                if p in used_dims:
+                    raise ValueError("pattern contains dimension {0} more "
+                                     "than once".format(p))
+                used_dims.add(p)
+            elif p == 'x':
+                # Broadcast
+                pass
+            else:
+                raise ValueError("pattern should only contain dimension"
+                                 "indices or 'x', not {0}".format(p))
+
+        self.pattern = pattern
+
+    def get_output_shape_for(self, input_shape):
+        # Build output shape while keeping track of the dimensions that we are
+        # attempting to collapse, so we can ensure that they are broadcastable
+        output_shape = []
+        dims_used = [False] * len(input_shape)
+        for p in self.pattern:
+            if isinstance(p, int):
+                if p < 0  or  p >= len(input_shape):
+                    raise ValueError("pattern contains {0}, but input shape has "
+                        "{1} dimensions only".format(p, len(input_shape)))
+                # Dimension p
+                o = input_shape[p]
+                dims_used[p] = True
+            elif p == 'x':
+                # Broadcast; will be of size 1
+                o = 1
+            else:
+                raise RuntimeError("invalid pattern entry, should have "
+                    "caught in the constructor")
+            output_shape.append(o)
+
+        for i, (dim_size, used) in enumerate(zip(input_shape, dims_used)):
+            if not used and dim_size != 1:
```

Comment:
`dim_size` could also be `None`. In this case, we cannot know whether it's okay to collapse it and should just assume that it is.